src/main/scala/org/spartanz/parserz/ParsersModule2.scala

```diff
@@ -0,0 +1,326 @@
+              unfold.lift(a) match {
+                case None          => (a, acc)
+                case Some((a1, b)) => rec(b :: acc)(a1)
+              }
+            rec(Nil)(arg)
+          }
+        )
+    }
+  }
+
+
+
+  private class ParserState[+E](val input: Input, var i: Int, var e: List[InOut.Error])
+
+  final def parser[S, E, A](grammar: Grammar[S, S, E, A]): Input => E \/ A =
+    input => step(grammar, new ParserState(input, 0, Nil))
+
+  private def step[S, E, A](grammar: Grammar[S, S, E, A], ps: ParserState[E]): E \/ A =
+    grammar match {
+      case Grammar.GADT.Consume(c, cond, e) =>
+        import InOut.Consumer.Simple._
+        val cas = c.feed(c.create(), ps.input, ps.i)
+        ps.i += extractCount(cas)
+        try {
+          val res = c.finish(extractState(cas))
+          if (cond == null || cond(res)) Right(res) else Left(e)
+        }
+        catch {
+          case InOut.NoInput  => Left(e)
+        }
+
+      case Grammar.GADT.ConsumeMany(c, e) =>
+        val state = c.create()
+        val count = c.feed(state, ps.input, ps.i)
+        ps.i += count
+        try {
+          Right(c.finish(state))
+        }
+        catch {
+          case InOut.NoInput  => Left(e)
+        }
+
+      case Grammar.GADT.ConsumeToken(c, e) =>
+        val state = c.create()
+        val count = c.feed(state, ps.input, ps.i)
+        ps.i += count
+        try {
+          val res = c.finish(state)
+          if (res == null) Left(e) else Right(res)
+        }
+        catch {
+          case InOut.NotFound => Left(e)
+          case InOut.NoInput  => Left(e)
+        }
+
+
+      case Grammar.GADT.Produce(a)     => Right(a)
+      case Grammar.GADT.Tag(value, _)  => step(value, ps)
+      case Grammar.GADT.Delay(delayed) => step(delayed(), ps)
+
+      case Grammar.GADT.Map(value, to, _) =>
+        step(value, ps).flatMap(to)
+
+      case Grammar.GADT.Filter(value, e, expr) =>
+        step(value, ps).flatMap {
+          a =>
+            if (exprFilter(expr)(a)) Right(a)
+            else Left(e)
+        }
+
+      case zip: Grammar.GADT.Zip[S, S, E, ta, tb] =>
+        val res1: E \/ ta = step(zip.left, ps)
+        val ret: E \/ (ta, tb) = res1 match {
+          case Left(e1) => Left(e1)
+          case Right(a) =>
+            val res2 = step(zip.right, ps)
+            res2.map[(ta, tb)](b => (a, b))
+        }
+        ret
+
+      case zip: Grammar.GADT.ZipL[S, S, E, ta, tb] =>
+        val res1: E \/ ta = step(zip.left, ps)
+        res1 match {
+          case Left(e1) => Left(e1)
+          case Right(a) =>
+            val res2 = step(zip.right, ps)
+            res2.map(_ => a)
+        }
+
+      case zip: Grammar.GADT.ZipR[S, S, E, ta, tb] =>
+        val res1: E \/ ta = step(zip.left, ps)
+        res1 match {
+          case Left(e1) => Left(e1)
+          case Right(_) => step(zip.right, ps)
+        }
+
+      case alt: Grammar.GADT.Alt[S, S, E, ta, tb] =>
```

Comment:
This should be replaced by `AltAll` so we can eliminate intermediate allocations (nested eithers).